### DISCUSSION TITLE
Exception handling in `Procedure`

### DISCUSSION THREAD

**Discussion by bicarlsen** [pymeasure/pymeasure#711] [Q&A]
Is there a canonical way to handle exceptions coming from a `Procedure` while begin run in a `Worker`?

e.g.
I have a `RuntimeError` that occurs in my procedure if certain measurement parameters are exceeded and would like to handle this error in the top-level program.

**Comment by BenediktBurger** [comment]
I'd handle errors in the procedure. E.g. handle an error setting/reading an instrument. I'm, however, no expert in the measurement part of pymeasure.

**Reply by bicarlsen** [reply]
I need to pass the error to the top-level program for other reasons (e.g. to inform the user a value was exceeded), so handling the error within the procedure is not possible.

**Reply by BenediktBurger** [reply]
You might open a message box from the procedure itself in order to inform the user.
Additionally, you could log the errors and the log handler does inform the user.

**Reply by bicarlsen** [reply]
Could you provide an example of how to do these? Especially logging the errors.

**Reply by BenediktBurger** [reply]
You just need to call `log.error` (or in an exception handler `log.exception`), at least if you follow the [default GUI example](https://pymeasure.readthedocs.io/en/latest/tutorial/graphical.html#using-the-managedwindow), which already has some logging in it. That way the exceptions are logged (and shown in the _experiment log_ pane) together with the other info messages.
If you want your log entries separate, you could add your own logger (`my_log = logging.getLogger("whatever")`) and add a handler in your main program to this logger. This handler may print the errors or save them etc. See the [logging documentation.](https://docs.python.org/3/library/logging.html)

**Reply by bicarlsen** [reply]
Ah. After looking at this, it actually doesn't cover my use case. and I would still need to propogate this information to the top-level program.

**Reply by BenediktBurger** [reply]
You could also use zmq and listen on the topic "error", because the worker publishes there the errors encountered during startup or execution of the procedure.
If you describe your use case in more detail or what you want to achieve, we might be in a better position to help you.

**Reply by bicarlsen** [reply]
After a quick look I think `zmq` will be the way to go. I'll give it a shot and come back with results.

**Use case**
We are building a [UI framework](https://github.com/EPFL-LPI/iv_lab/tree/lab) using `pymeasure` as a basis for some core functionality (I'm planning to share here in the `Show and tell` topic when it's complete). We were having a problem debugging a Procedure because it is running in a separate thread, so the measurement would just die, but we didn't have a way to get at the exception or stack trace to see why. This made the debugging process quite slow and tedious, requiring us to use `print` statements to narrow down the cause. I was hoping there is a way to get access to these unhandled exceptions to make this process easier in the future.

**Reply by bilderbuchi** [reply]
I think using the existing zmq infrastructure is a good idea, it's a natural fit for these kind of communication needs.
If there are small tweaks on our side to make this a better fit for your use case, we can talk about that.

**Comment by msmttchr** [comment]
You could check also https://docs.python.org/3/library/threading.html#threading.excepthook

**Reply by bicarlsen** [reply]
I briefly tried implementing an `excepthook` with a logger, but couldn't get any output. This very possible could have been an error on my side. If you have an example that works to post it would be greatly appreciated.

**Reply by msmttchr** [reply]
I have extracted this part from working code, bit I   am not sure everything is included.
Please note, the working code is using threading but outside pymeasure.
```
import sys, threading

def UnhandledException(e_type, value, tb):
        traceback_strs = traceback.format_exception(e_type, value, tb)
        exception_str = "".join(traceback_strs)
        print (exception_str)

sys.excepthook = UnhandledException

def thread_exception_handler(**kwargs):
   print("thread_exception_handler",kwargs)

threading.excepthook = thread_exception_handler
```
